### DISCUSSION TITLE
Best Replacement for `dict.clear()` in a Typed Dict?

### DISCUSSION THREAD

**Discussion by ImplyingICheck** [microsoft/pyright#5730] [Q&A] [answered]
### Problem
In a project I am working on, I have to normalize two nested dictionaries to compare their contents. Part of this process is replacing a volatile value with a "dummy" variable consisting of a dictionary with sentinel values.

### Minimalist Example
```python
from typing import NotRequired, TypedDict


class Foo(TypedDict):
  bar: str
  volatile_values: NotRequired[str]


def clear_and_update(original_data: Foo, data_to_copy: Foo):
  original_data.clear()
  original_data.update(data_to_copy)


# Equivalent effect:
def clear_original_set_minus_copy_and_update(
    original_data: Foo, data_to_copy: Foo
):
  for key in list(original_data):
    if key not in data_to_copy:
      del original_data[key]
  original_data.update(data_to_copy)


if __name__ == '__main__':
  new_data = Foo(bar='I replace Hello World!')
  original_clear = Foo(bar='Hello world!', volatile_values='Oof')
  clear_and_update(original_clear, new_data)
  original_set_minus = Foo(bar='Hello world!', volatile_values='Bar')
  clear_original_set_minus_copy_and_update(original_set_minus, new_data)
  # Check that each dictionary is a deep copy of new_data
  assert original_clear == original_set_minus
  assert original_clear is not original_set_minus

```
While `clear_and_update()` does not have a runtime error, `pyright` correctly reports that `dict.clear()` is not a member of `TypedDict`. Is there an intended way to replicate the `dict.clear()` call in a type-safe manner?

Ideally, the solution would not require significant auxiliary storage, similar to the implementation of `dict.clear()` in CPython.

**Comment by erictraut** [comment, marked as answer]
> Is there an intended way to replicate the `dict.clear()` call in a type-safe manner?

No, I don't think so. Calling `clear` on a TypedDict isn't type safe. This is true even in cases where all of the fields are `NotRequired` (which isn't the case in your code sample) because there may be other fields that are required in subclasses.
